### ISSUE TITLE
FAISS score filtering is done twice with and without applied relevance_fn

### ISSUE THREAD

**Issue by ozgurdemir** [langchain-ai/langchain#21419]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

`FAISS._similarity_search_with_score_by_vector` includes the following code:

```
...
if score_threshold is not None:
            cmp = (
                operator.ge
                if self.distance_strategy
                in (DistanceStrategy.MAX_INNER_PRODUCT, DistanceStrategy.JACCARD)
                else operator.le
            )
            docs = [
                (doc, similarity)
                for doc, similarity in docs
                if cmp(similarity, score_threshold)
            ]
...
docs_and_rel_scores = [
            (doc, relevance_score_fn(score)) for doc, score in docs_and_scores
        ]
...

```

in other words the entries are filtered by score. This happens **before** the `relevance_fn` has been applied.

However, after this first filtering step the filtering happens again in `VectorStore.similarity_search_with_relevance_scores` this time **after** the relevancy_fn has been applied. 

```
...
docs_and_similarities = self._similarity_search_with_relevance_scores(
...
if score_threshold is not None:
            docs_and_similarities = [
                (doc, similarity)
                for doc, similarity in docs_and_similarities
                if similarity >= score_threshold
            ]
            if len(docs_and_similarities) == 0:
                warnings.warn(
                    "No relevant docs were retrieved using the relevance score"
                    f" threshold {score_threshold}"
                )
        return docs_and_similarities
...
```

### Description

Double filtering with and without relevance_fn does not allow to set an appropiate score threshold.

### System Info

langchain                        0.1.17                                            
langchain-community              0.0.37                                
langchain-core                   0.1.52                                       
langchain-openai                 0.0.6                                                                                                                       
langchain-text-splitters         0.0.1

**Comment by habanoz** [comment]
@ozgurdemir I have been debugging sth. else and found your issue here. A quick question. Are you sure the filtering is applied twice?

The first line in "VectorStore.similarity_search_with_relevance_scores" removes "score_threshold".
score_threshold = kwargs.pop("score_threshold", None)

This should prevent "FAISS.similarity_search_with_score_by_vector" (no preceding underscore here) from applying filtering...

Just wanted to help.
Regards

**Comment by ozgurdemir** [comment]
Hey @habanoz ,

thx for your reply. I'm not working on this particular project anymore. I remember that we stumbled on this as we had a weird behaviour when using the `relevance_score_fn`.

However, looking at the code again it seems like you're right.
`VectorStoreRetreiver` calls the `VectorStore.similarity_search_with_relevance_scores` which removes the `score_threshold` parameter and then calls `_similarity_search_with_relevance_scores` without this parameter. So the filtering step should be circumvented.

In other words the `relevance_score_fn` is applied to the returned scores but the filtering happens on the raw scores. 
All in all it seems like a very confusing implementation and probably not what is intended. The filtering by score depends on the `similarity_metric` used. So it should be either smaller_than or greater_than depending on the similarity_metric but this code path is not executed due to the above mentioned logic.

So I still think there's a bug here :(
